### ISSUE TITLE
Support setting host header when using JSON API

### ISSUE THREAD

**Issue by vjeffz** [GoogleCloudPlatform/gsutil#727]
When trying to use gsutil with [ VPC Service Controls](https://cloud.google.com/vpc-service-controls/) enabled, setting gs_json_host to the restricted.googleapis.com VIP (199.36.153.4/30) throws a 404. 

```
BucketNotFoundException: 404 gs://<<bucket>> bucket does not exist.
```

This is because the Host header is set to 'restricted.googleapis.com' and the VIP doesn't forward the request to the correct API. 

```
...
send: 'GET /storage/v1/b/<bucket>/o?delimiter=%2F&projection=noAcl&versions=False&fields=prefixes%2CnextPageToken%2Citems%2Fname&alt=json&maxResults=1000 HTTP/1.1\r\nHost: restricted.googleapis.com\r\ncontent-length: 0\r\nauthorization: Bearer xxx\r\naccept-encoding: gzip, deflate\r\naccept: application/json\r\nuser-agent: apitools gsutil/4.37 Python/2.7.15 (win32) google-cloud-sdk/240.0.0 analytics/disabled\r\n\r\n'
reply: 'HTTP/1.1 404 Not Found\r\n'
header: Content-Type: text/html; charset=UTF-8
header: Referrer-Policy: no-referrer
header: Content-Length: 1598
header: Date: Fri, 05 Apr 2019 03:43:36 GMT
...
<p>The requested URL <code>/storage/v1/b/<bucket>/o</code> was not found on this server.  <ins>That\ufffd\ufffd\ufffds all we know.</ins>
```

To get this working locally I patched gcs_json_api.py to support `gs_json_host_header` in the .boto config which appends the specified host to `additional_http_headers` . Now when I run the following, the host header is set correctly.

```
gsutil -DD -o  "Credentials:gs_json_host_header=www.googleapis.com" -o "Credentials:gs_json_host=restricted.googleapis.com" ls gs://<bucket>/
```

Output:
```
send: 'GET /storage/v1/b/<bucket>/o?delimiter=%2F&projection=noAcl&versions=False&fields=prefixes%2CnextPageToken%2Citems%2Fname&alt=json&maxResults=1000 HTTP/1.1\r\ncontent-length: 0\r\naccept-encoding: gzip, deflate\r\naccept: application/json\r\nuser-agent: apitools gsutil/4.37 Python/2.7.15 (win32) google-cloud-sdk/240.0.0 analytics/disabled\r\nhost: www.googleapis.com\r\nauthorization: Bearer xxx\r\n\r\n'
reply: 'HTTP/1.1 200 OK\r\n'
header: X-GUploader-UploadID: xxx
header: Vary: Origin
header: Vary: X-Origin
header: Content-Type: application/json; charset=UTF-8
header: Expires: Fri, 05 Apr 2019 03:38:53 GMT
header: Date: Fri, 05 Apr 2019 03:38:53 GMT
header: Cache-Control: private, max-age=0, must-revalidate, no-transform
header: Content-Length: 46
header: Server: UploadServer
INFO 0405 14:38:51.512000 base_api.py] Response of type Objects: <Objects
 items: []
 prefixes: [u'test1/', u'test2/']>
```

I'm in the process of following the contribution guide to get the code submitted for review.

**Comment by houglum** [comment]
Thanks for filing an issue! We're in the process of revising the contribution/dev instructions -- you can just make a fork of this repo and submit a PR on GitHub.

I'm not quite sure how/if a request is *supposed* to be changed to work with VPC - the docs are a bit long-winded and seem to say different things in different places.  The docs at [1] mention:

> your VM instances must send requests to restricted.googleapis.com instead of *.googleapis.com

(which seems to be what you're doing above), but then go on to recommend you configure your DNS server to accomplish this instead, i.e.:

> "To make use of the Restricted Google APIs IP addresses, configure your DNS server to resolve *.googleapis.com as a CNAME to restricted.googleapis.com

It seems like the latter would take care of this -- you send the request to www.googleapis.com, the host would be set to www.googleapis.com, and the request would get routed to the VPC VIP, unless I'm overlooking something.

I can't find any examples of this offhand, but I'll try to poke someone tomorrow who can show me how this is all supposed to be set up and how VPC requests should be formatted to reach Cloud Storage.

[1] https://cloud.google.com/vpc-service-controls/docs/set-up-private-connectivity

**Comment by vjeffz** [comment]
Ah right, I'll submit a PR when I get a chance.  Yes modifying DNS does work. There's 2 issues with this:

1) We are accessing the network vpc  via on-prem through Partner Interconnect. The server that is running gsutil cannot easily update their DNS servers or /etc/hosts.  

2)  Using the Storage API specific endpoints `storage-component.googleapis.com` and  `storage.googleapis.com` doesn't seem to work with gsutil. Modifying DNS to CNAME *.googleapis.com to restricted.googleapis.com will route ALL Google API traffic from on-premise through the interconnect which is not something we're willing to do right now.  

With the suggested fix, we can isolate specific applications to either use the private restricted.googleapis.com VIP (if their GCS resource is protected with VPC Service controls) or continue to use the www.googleapis.com endpoint which routes to public internet by default.

**Comment by houglum** [comment]
Ah, thanks for the explanation!

It's worth noting the gsutil also makes requests to other Cloud JSON APIs, i.e. PubSub at pubsub.googleapis.com and KMS at cloudkms.googleapis.com (all 3 of our apitools-generated clients can be found in `gslib/third_party/`, and we have corresponding wrappers around them in `gslib/`, e.g. `kms_api.py` and `pubsub_api.py`).  I don't think we ever got around to putting the KMS and PubSub host attributes in users' generated boto files, but the similarly named attributes are `gs_pubsub_host` and `gs_kms_host`.

Also, OAuth2 tokens are fetched from oauth2.googleapis.com.  All this to say: we send requests to a lot of subdomains under `*.googleapis.com`, and it seems odd to only add this host-header functionality to *some* of our `<service>_api` modules.

While my initial reaction is to say that we should add your suggested functionality to *all* our client libraries, I'm fine with not adding those bits until somebody asks for them.  As far as I understand, it's fairly uncommon to send some traffic (i.e. only GCS requests) over VPC and other traffic over the public internet, so I don't anticipate that users will need to be able to set the other host-header attributes.

TL;DR: We might have to do the same thing for other services gsutil makes requests to in the future, but let's go ahead and just add this functionality for the `gcs_json_api` module for now.

**Comment by vjeffz** [comment]
It really depends on the Service Perimeter configuration and the restricted services you set. (https://cloud.google.com/vpc-service-controls/docs/supported-products#supported_products).  If those APIs such as Pub/Sub and KMS are part of the perimeter then yes you'll need to implement this functionality for those modules as well.
